### PR TITLE
Type and provider refactor

### PR DESCRIPTION
**Pull request by petems**
* Refactor swap_file to use type and provider
* Deprecate usage of class to create swap_files (will remove in next major release)
* Add defined type to use instead of class

### CODE REVIEW

File: lib/puppet/provider/swap_file/linux.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    instances.each do |prov|
+      if resource = resources[prov.name]
+        resource.provider = prov
+      end
+    end
+  end
+
+  def self.instances
+    get_swap_files.collect do |swapfile_line|
+      swapfile_properties = get_swapfile_properties(swapfile_line)
```
Comment (garethr) [inline lib/puppet/provider/swap_file/linux.rb]: You don't really need the intermediary var here `new(get_swapfile_properties(swapfile_line))`

File: lib/puppet/provider/swap_file/linux.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    swapfile_properties = {}
+
+    # swapon -s output formats thus:
+    # Filename        Type    Size  Used  Priority
+
+    # Split on spaces
+    output_array = swapfile_line.strip.split(/\s+/)
+
+    # Assign properties based on headers
+    swapfile_properties[:ensure]     = :present
```
Comment (garethr) [inline lib/puppet/provider/swap_file/linux.rb]: I'd just write this out as a hash and save on the typing of swapfile_properties:

```
swapfile_properties = {
  name: output_array[0].
  ...
}
```

Bonus bonus if you just have an array of keys given the progression on the value side.

**Comment by garethr** [comment]
If I was being pedantic (never) I'd say improve the type with better input validation and requisite tests. But other that the passing comments on a very brief run through looks OK.

**Comment by petems** [comment]
@garethr  Good plan, I need to actually read the swapfile man pages and find out what some sensible validation is.

Thank's the review! :+1:

File: lib/puppet/provider/swap_file/linux.rb (hunk 2 above)
Comment (petems) [inline lib/puppet/provider/swap_file/linux.rb]: > Bonus bonus if you just have an array of keys given the progression on the value side.

Can you clarify on this bit?

Right now I've just refactored it to:

```ruby
swapfile_properties = {
      ensure: :present,
      name: output_array[0],
      file: output_array[0],
      type: output_array[1],
      size: output_array[2],
      used: output_array[3],
      priority: output_array[4]
    }
```

Comment (garethr) [inline lib/puppet/provider/swap_file/linux.rb]: That works. You could probably get rid of the repetition of output_array if you really tried, by iterating over an array of the keys. But it's purely a style thing.

**Comment by petems** [comment]
Ok, so the only thing I can validate is the parameter of file, which I've added in and added tests :+1:

**Pull request merged**